src/Lecture2.hs

```diff
@@ -207,9 +323,34 @@ The algorithm of merge sort is the following:
 >>> mergeSort [3, 1, 2]
 [1,2,3]
 -}
+
+-- First take.
+-- I think there should be a way to not to use `length` here
+-- but I couldn't find one yet.
+--mergeSort0 :: Ord a => [a] -> [a]
+mergeSort0 :: [Int] -> [Int]
+mergeSort0 [] = []
+mergeSort0 [x] = [x]
+mergeSort0 xs = merge (mergeSort0 left) (mergeSort0 right)
+    where (left, right) = splitAt (length xs `div` 2) xs
```

Comment:
You can avoid using `length` if you notice that you don't necessarily need to split in the middle. You only need to split it into two lists of approximately the same size.

In that case, you can implement a solution that traverses a list only once and splits its elements into two lists of elements on only even positions and only odd positions.

Reply:
Found this trick: https://github.com/quchen/articles/blob/master/2018-11-22_zipWith_const.md#example-2-split-list-in-half
But it's to complicated for now, so resolving with:
```Haskell
mergeSort :: [Int] -> [Int]
mergeSort [] = []
mergeSort [x] = [x]
mergeSort xs = uncurry merge . bimap mergeSort mergeSort $ split2 [] [] xs
    where split2 h1 h2 [] = (h1, h2)
          split2 h1 h2 [x] = (h1, x:h2)
          split2 h1 h2 (x:y:xs') = split2 (x:h1) (y:h2) xs'
```